### ISSUE TITLE
New Solr plugin to collect stats via metrics API

### ISSUE THREAD

**Issue by frizner** [influxdata/telegraf#5907]
## Feature Request
Since Solr 6.4,  huge variety of Solr, JVM, system related stats can be gather by usage of [Metrics API](https://lucene.apache.org/solr/guide/metrics-reporting.html#metrics-api).

### Proposal:
A new `solrmetrics` plugin will allow gather stats via the API.
Gathering needed Solr related stats, including JVM, Jetty, Solr cores, replication etc, by setting only needed keys and prefix for `/solr/admin/metrics` endpoint of Solr.

### Use case: 
Setting prefixes and keys for Solr metrics in the plugin lets retrieve  full and only needed stats  for monitoring of different scenarios of Solr usage.

PR - https://github.com/influxdata/telegraf/pull/5908

**Comment by frizner** [comment]
https://github.com/influxdata/telegraf/pull/5908

**Comment by kgelinas** [comment]
Which version of the Metrics API will this be compatible?
will this be compatible with 6.x branch
https://lucene.apache.org/solr/guide/6_6/metrics-reporting.html#metrics-api

**Comment by frizner** [comment]
The plugin should work with Solr 6.x branch too. In this case an user shouldn't set keys. Otherwise Solr returns all accessible metrics.

**Comment by kgelinas** [comment]
I've tried with a simple prefixes
- solr.jetty:org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses does not work
- org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses
  return this:
    2019-07-10T17:06:30Z E! [inputs.solrmetrics]: Error in plugin: json: cannot unmarshal array into Go 
struct field MetricResponse.metrics of type solrmetrics.Metrics

probably because there's some extra solr.core. solr.jvm solr.node with the output
if I only want solr.jetty I have to specify a group
solr/admin/metrics?compact=true&prefix=org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses&**group**=solr.jetty&wt=json

**Comment by frizner** [comment]
> probably because there's some extra solr.core. solr.jvm solr.node with the output
correct. It is hard to parse such output and set right tags and values for the metrics. It is why in this case `keys` should be used, but it is not your solution because of Solr 6. Right?

**Comment by kgelinas** [comment]
exactly
keys are not present in 6.x

**Comment by kgelinas** [comment]
I don't think there's an easy way to detect solr version if there is maybe it's a better way.
if not you could add a flag for 6x compatibility
reuse the same prefix format "solr.jetty:org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses" 
if the flag is present modify the url to put the value before the ":" to group in the url?

**Comment by frizner** [comment]
Initially I wanted to use groups, but it didn't looks good solution, because groups can be combined with prefixes and the config becomes too confusing to handle or understand.

**Comment by frizner** [comment]
@kgelinas 
Could you show output of `http://solrserver:8389/solr/admin/metrics?wt=json&prefix=org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses` and your config file?
I tested the plugin with Solr 7 with only one prefix and it works. 
```
prefixes = ["org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses"]
```
I see the following metrics in a DB with common prefix `solr.jetty.org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses`:
"count", "meanRate", "1minRate", "5minRate","15minRate"

**Comment by kgelinas** [comment]
{"responseHeader":{"status":0,"QTime":4},"metrics":["solr.core.customers",{},"solr.core.xxxx",{},"solr.core.xxxxxx",{},"solr.jetty",{"org.eclipse.jetty.server.handler.DefaultHandler.5xx-responses":{"count":1,"meanRate":1.7572203828424237E-6,"1minRate":2.964393875E-314,"5minRate":7.595114699769402E-299,"15minRate":3.1499831990439436E-102}},"solr.core.agencyDetail",{},"solr.jvm",{},"solr.node",{},"solr.core.xxxxxx",{},"solr.core.xxxx",{}]}

**Comment by frizner** [comment]
Solr 6.x returns metrics as an array, Solr 7.x as a map.

**Comment by sjwang90** [comment]
Closing issue due to inactivity and interested. If someone is interested on taking this work back up, please take a look at PR #5908.
